README.md
# DeployM365Apps

Deploy the Microsoft 365 Apps for Enterpirse via Microsoft CDN with PowerShell Application Deployment Toolkit 3.8.3.


CHANGELOG
# CHANGELOG

2022-03-06 - Update PSADT to `3.8.4` from `3.8.3`.
             Update Update Channel to `Current` from `SemiAnnual`.

2021-01-17 - Update to new repository.
             Update scripts to PowerShell Application Deployment Toolkit 3.8.3 from 3.8.2.


AppDeployToolkit/AppDeployToolkitHelp.ps1
<#
.SYNOPSIS
  Displays a graphical console to browse the help for the App Deployment Toolkit functions
    # LICENSE #
    PowerShell App Deployment Toolkit - Provides a set of functions to perform common application deployment tasks on Windows.
    Copyright (C) 2017 - Sean Lillis, Dan Cunningham, Muhammad Mashwani, Aman Motazedian.
    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
.DESCRIPTION
  Displays a graphical console to browse the help for the App Deployment Toolkit functions
.EXAMPLE
  AppDeployToolkitHelp.ps1
.NOTES
.LINK
  http://psappdeploytoolkit.com
#>

##*===============================================
##* VARIABLE DECLARATION
##*===============================================

## Variables: Script
[string]$appDeployToolkitHelpName = 'PSAppDeployToolkitHelp'
[string]$appDeployHelpScriptFriendlyName = 'App Deploy Toolkit Help'
[version]$appDeployHelpScriptVersion = [version]'3.8.3'
[string]$appDeployHelpScriptDate = '30/09/2020'

## Variables: Environment
[string]$scriptDirectory = Split-Path -Path $MyInvocation.MyCommand.Definition -Parent
#  Dot source the App Deploy Toolkit Functions
. "$scriptDirectory\AppDeployToolkitMain.ps1" -DisableLogging
. "$scriptDirectory\AppDeployToolkitExtensions.ps1"
##*===============================================
##* END VARIABLE DECLARATION
##*===============================================

##*===============================================
##* FUNCTION LISTINGS
##*===============================================

Function Show-HelpConsole {
  ## Import the Assemblies
  Add-Type -AssemblyName 'System.Windows.Forms' -ErrorAction 'Stop'
  Add-Type -AssemblyName System.Drawing -ErrorAction 'Stop'

  ## Form Objects
  $HelpForm = New-Object -TypeName 'System.Windows.Forms.Form'
  $HelpListBox = New-Object -TypeName 'System.Windows.Forms.ListBox'
  $HelpTextBox = New-Object -TypeName 'System.Windows.Forms.RichTextBox'
  $InitialFormWindowState = New-Object -TypeName 'System.Windows.Forms.FormWindowState'

  ## Form Code
  $System_Drawing_Size = New-Object -TypeName 'System.Drawing.Size'
  $System_Drawing_Size.Height = 665
  $System_Drawing_Size.Width = 957
  $HelpForm.ClientSize = $System_Drawing_Size
  $HelpForm.DataBindings.DefaultDataSourceUpdateMode = 0
  $HelpForm.Name = 'HelpForm'
  $HelpForm.Text = 'PowerShell App Deployment Toolkit Help Console'
  $HelpForm.WindowState = 'Normal'
  $HelpForm.ShowInTaskbar = $true
  $HelpForm.FormBorderStyle = 'Fixed3D'
  $HelpForm.MaximizeBox = $false
  $HelpForm.Icon = New-Object -TypeName 'System.Drawing.Icon' -ArgumentList $AppDeployLogoIcon
  $HelpListBox.Anchor = 7
  $HelpListBox.BorderStyle = 1
  $HelpListBox.DataBindings.DefaultDataSourceUpdateMode = 0
  $HelpListBox.Font = New-Object -TypeName 'System.Drawing.Font' -ArgumentList ('Microsoft Sans Serif', 9.75, 1, 3, 1)
  $HelpListBox.FormattingEnabled = $true
  $HelpListBox.ItemHeight = 16
  $System_Drawing_Point = New-Object -TypeName 'System.Drawing.Point'
  $System_Drawing_Point.X = 0
  $System_Drawing_Point.Y = 0
  $HelpListBox.Location = $System_Drawing_Point
  $HelpListBox.Name = 'HelpListBox'
  $System_Drawing_Size = New-Object -TypeName 'System.Drawing.Size'
  $System_Drawing_Size.Height = 658
  $System_Drawing_Size.Width = 271
  $HelpListBox.Size = $System_Drawing_Size
  $HelpListBox.Sorted = $true
  $HelpListBox.TabIndex = 2
  $HelpListBox.add_SelectedIndexChanged({ $HelpTextBox.Text = Get-Help -Name $HelpListBox.SelectedItem -Full | Out-String })
  $helpFunctions = Get-Command -CommandType 'Function' | Where-Object { ($_.HelpUri -match 'psappdeploytoolkit') -and ($_.Definition -notmatch 'internal script function') } | Select-Object -ExpandProperty Name
  $null = $HelpListBox.Items.AddRange($helpFunctions)
  $HelpForm.Controls.Add($HelpListBox)
  $HelpTextBox.Anchor = 11
  $HelpTextBox.BorderStyle = 1
  $HelpTextBox.DataBindings.DefaultDataSourceUpdateMode = 0
  $HelpTextBox.Font = New-Object -TypeName 'System.Drawing.Font' -ArgumentList ('Microsoft Sans Serif', 8.5, 0, 3, 1)
  $HelpTextBox.ForeColor = [System.Drawing.Color]::FromArgb(255, 0, 0, 0)
  $System_Drawing_Point = New-Object -TypeName System.Drawing.Point
  $System_Drawing_Point.X = 277
  $System_Drawing_Point.Y = 0
  $HelpTextBox.Location = $System_Drawing_Point
  $HelpTextBox.Name = 'HelpTextBox'
  $HelpTextBox.ReadOnly = $True
  $System_Drawing_Size = New-Object -TypeName 'System.Drawing.Size'
  $System_Drawing_Size.Height = 658
  $System_Drawing_Size.Width = 680
  $HelpTextBox.Size = $System_Drawing_Size
  $HelpTextBox.TabIndex = 1
  $HelpTextBox.Text = ''
  $HelpForm.Controls.Add($HelpTextBox)

  ## Save the initial state of the form
  $InitialFormWindowState = $HelpForm.WindowState
  ## Init the OnLoad event to correct the initial state of the form
  $HelpForm.add_Load($OnLoadForm_StateCorrection)
  ## Show the Form
  $null = $HelpForm.ShowDialog()
}

##*===============================================
##* END FUNCTION LISTINGS
##*===============================================

##*===============================================
##* SCRIPT BODY
##*===============================================

Write-Log -Message "Load [$appDeployHelpScriptFriendlyName] console..." -Source $appDeployToolkitHelpName

## Show the help console
Show-HelpConsole

Write-Log -Message "[$appDeployHelpScriptFriendlyName] console closed." -Source $appDeployToolkitHelpName

##*===============================================
##* END SCRIPT BODY
##*===============================================


AppDeployToolkit/AppDeployToolkitExtensions.ps1
<#
.SYNOPSIS
  This script is a template that allows you to extend the toolkit with your own custom functions.
    # LICENSE #
    PowerShell App Deployment Toolkit - Provides a set of functions to perform common application deployment tasks on Windows.
    Copyright (C) 2017 - Sean Lillis, Dan Cunningham, Muhammad Mashwani, Aman Motazedian.
    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
.DESCRIPTION
  The script is automatically dot-sourced by the AppDeployToolkitMain.ps1 script.
.NOTES
    Toolkit Exit Code Ranges:
    60000 - 68999: Reserved for built-in exit codes in Deploy-Application.ps1, Deploy-Application.exe, and AppDeployToolkitMain.ps1
    69000 - 69999: Recommended for user customized exit codes in Deploy-Application.ps1
    70000 - 79999: Recommended for user customized exit codes in AppDeployToolkitExtensions.ps1
.LINK
  http://psappdeploytoolkit.com
#>
[CmdletBinding()]
Param (
)

##*===============================================
##* VARIABLE DECLARATION
##*===============================================

# Variables: Script
[string]$appDeployToolkitExtName = 'PSAppDeployToolkitExt'
[string]$appDeployExtScriptFriendlyName = 'App Deploy Toolkit Extensions'
[version]$appDeployExtScriptVersion = [version]'3.8.3'
[string]$appDeployExtScriptDate = '30/09/2020'
[hashtable]$appDeployExtScriptParameters = $PSBoundParameters

##*===============================================
##* FUNCTION LISTINGS
##*===============================================

# <Your custom functions go here>

##*===============================================
##* END FUNCTION LISTINGS
##*===============================================

##*===============================================
##* SCRIPT BODY
##*===============================================

If ($scriptParentPath) {
  Write-Log -Message "Script [$($MyInvocation.MyCommand.Definition)] dot-source invoked by [$(((Get-Variable -Name MyInvocation).Value).ScriptName)]" -Source $appDeployToolkitExtName
} Else {
  Write-Log -Message "Script [$($MyInvocation.MyCommand.Definition)] invoked directly" -Source $appDeployToolkitExtName
}

##*===============================================
##* END SCRIPT BODY
##*===============================================


Deploy-Application.ps1
<#
.SYNOPSIS
  This script performs the installation or uninstallation of an application(s).
  # LICENSE #
  PowerShell App Deployment Toolkit - Provides a set of functions to perform common application deployment tasks on Windows.
  Copyright (C) 2017 - Sean Lillis, Dan Cunningham, Muhammad Mashwani, Aman Motazedian.
  This program is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
  You should have received a copy of the GNU Lesser General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
.DESCRIPTION
  The script is provided as a template to perform an install or uninstall of an application(s).
  The script either performs an "Install" deployment type or an "Uninstall" deployment type.
  The install deployment type is broken down into 3 main sections/phases: Pre-Install, Install, and Post-Install.
  The script dot-sources the AppDeployToolkitMain.ps1 script which contains the logic and functions required to install or uninstall an application.
.PARAMETER DeploymentType
  The type of deployment to perform. Default is: Install.
.PARAMETER DeployMode
  Specifies whether the installation should be run in Interactive, Silent, or NonInteractive mode. Default is: Interactive. Options: Interactive = Shows dialogs, Silent = No dialogs, NonInteractive = Very silent, i.e. no blocking apps. NonInteractive mode is automatically set if it is detected that the process is not user interactive.
.PARAMETER AllowRebootPassThru
  Allows the 3010 return code (requires restart) to be passed back to the parent process (e.g. SCCM) if detected from an installation. If 3010 is passed back to SCCM, a reboot prompt will be triggered.
.PARAMETER TerminalServerMode
  Changes to "user install mode" and back to "user execute mode" for installing/uninstalling applications for Remote Destkop Session Hosts/Citrix servers.
.PARAMETER DisableLogging
  Disables logging to file for the script. Default is: $false.
.EXAMPLE
    powershell.exe -Command "& { & '.\Deploy-Application.ps1' -DeployMode 'Silent'; Exit $LastExitCode }"
.EXAMPLE
    powershell.exe -Command "& { & '.\Deploy-Application.ps1' -AllowRebootPassThru; Exit $LastExitCode }"
.EXAMPLE
    powershell.exe -Command "& { & '.\Deploy-Application.ps1' -DeploymentType 'Uninstall'; Exit $LastExitCode }"
.EXAMPLE
    Deploy-Application.exe -DeploymentType "Install" -DeployMode "Silent"
.NOTES
  Toolkit Exit Code Ranges:
  60000 - 68999: Reserved for built-in exit codes in Deploy-Application.ps1, Deploy-Application.exe, and AppDeployToolkitMain.ps1
  69000 - 69999: Recommended for user customized exit codes in Deploy-Application.ps1
  70000 - 79999: Recommended for user customized exit codes in AppDeployToolkitExtensions.ps1
.LINK
  http://psappdeploytoolkit.com
#>
[CmdletBinding()]
Param (
  [Parameter(Mandatory=$false)]
  [ValidateSet('Install','Uninstall','Repair')]
  [string]$DeploymentType = 'Install',
  [Parameter(Mandatory=$false)]
  [ValidateSet('Interactive','Silent','NonInteractive')]
  [string]$DeployMode = 'Interactive',
  [Parameter(Mandatory=$false)]
  [switch]$AllowRebootPassThru = $false,
  [Parameter(Mandatory=$false)]
  [switch]$TerminalServerMode = $false,
  [Parameter(Mandatory=$false)]
  [switch]$DisableLogging = $false
)

Try {
  ## Set the script execution policy for this process
  Try { Set-ExecutionPolicy -ExecutionPolicy 'ByPass' -Scope 'Process' -Force -ErrorAction 'Stop' } Catch {}

  ##*===============================================
  ##* VARIABLE DECLARATION
  ##*===============================================
  ## Variables: Application
  [string]$appVendor = 'Microsoft Corporation'
  [string]$appName = '365 Apps for Enterprise'
  [string]$appVersion = '' # No need to display this!
  [string]$appArch = 'x64'
  [string]$appLang = 'EN'
  [string]$appRevision = '01'
  [string]$appScriptVersion = '1.2.0'
  [string]$appScriptDate = '2021-01-17'
  [string]$appScriptAuthor = 'Cameron Kollwitz (Original by Sandy Zeng)'
  ##*===============================================
  ## Variables: Install Titles (Only set here to override defaults set by the toolkit)
  [string]$installName = ''
  [string]$installTitle = ''

  ##* Do not modify section below
  #region DoNotModify

  ## Variables: Exit Code
  [int32]$mainExitCode = 0

  ## Variables: Script
  [string]$deployAppScriptFriendlyName = 'Deploy Application'
  [version]$deployAppScriptVersion = [version]'3.8.3'
  [string]$deployAppScriptDate = '30/09/2020'
  [hashtable]$deployAppScriptParameters = $psBoundParameters

  ## Variables: Environment
  If (Test-Path -LiteralPath 'variable:HostInvocation') { $InvocationInfo = $HostInvocation } Else { $InvocationInfo = $MyInvocation }
  [string]$scriptDirectory = Split-Path -Path $InvocationInfo.MyCommand.Definition -Parent

  ## Dot source the required App Deploy Toolkit Functions
  Try {
    [string]$moduleAppDeployToolkitMain = "$scriptDirectory\AppDeployToolkit\AppDeployToolkitMain.ps1"
    If (-not (Test-Path -LiteralPath $moduleAppDeployToolkitMain -PathType 'Leaf')) { Throw "Module does not exist at the specified location [$moduleAppDeployToolkitMain]." }
    If ($DisableLogging) { . $moduleAppDeployToolkitMain -DisableLogging } Else { . $moduleAppDeployToolkitMain }
  }
  Catch {
    If ($mainExitCode -eq 0){ [int32]$mainExitCode = 60008 }
    Write-Error -Message "Module [$moduleAppDeployToolkitMain] failed to load: `n$($_.Exception.Message)`n `n$($_.InvocationInfo.PositionMessage)" -ErrorAction 'Continue'
    ## Exit the script, returning the exit code to SCCM
    If (Test-Path -LiteralPath 'variable:HostInvocation') { $script:ExitCode = $mainExitCode; Exit } Else { Exit $mainExitCode }
  }

  #endregion
  ##* Do not modify section above
  ##*===============================================
  ##* END VARIABLE DECLARATION
  ##*===============================================

  If ($deploymentType -ine 'Uninstall' -and $deploymentType -ine 'Repair') {
    ##*===============================================
    ##* PRE-INSTALLATION
    ##*===============================================
    [string]$installPhase = 'Pre-Installation'

    ## Show Welcome Message, close required applications, allow up to 3 deferrals, verify there is enough disk space to complete the install, and persist the prompt
    Show-InstallationWelcome -CloseApps "MSACCESS,EXCEL,INFOPATH,ONENOTEM,GROOVE,ONENOTE,OUTLOOK,POWERPNT,WINPROJ,MSPUB,SPDESIGN,lync,VISIO,WINWORD,Teams,Onedrive" -AllowDeferCloseApps -AllowDefer -DeferDays "1" -CloseAppsCountdown "5400" -PersistPrompt -BlockExecution

    ## Show Progress Message (with the default message)
    Show-InstallationProgress -StatusMessage "We are installing $installTitle. Please wait." -WindowLocation 'BottomRight' -TopMost $false

    ## <Perform Pre-Installation tasks here>


    ##*===============================================
    ##* INSTALLATION
    ##*===============================================
    [string]$installPhase = 'Installation'

    ## Handle Zero-Config MSI Installations
    If ($useDefaultMsi) {
      [hashtable]$ExecuteDefaultMSISplat =  @{ Action = 'Install'; Path = $defaultMsiFile }; If ($defaultMstFile) { $ExecuteDefaultMSISplat.Add('Transform', $defaultMstFile) }
      Execute-MSI @ExecuteDefaultMSISplat; If ($defaultMspFiles) { $defaultMspFiles | ForEach-Object { Execute-MSI -Action 'Patch' -Path $_ } }
    }

    ## <Perform Installation tasks here>
    Execute-Process -Path "$dirFiles\setup.exe" -WaitForMsiExec "/configure `"$dirSupportFiles\Install.xml`""

    ## Force update group policy
    Invoke-GPUpdate -RandomDelayInMinutes 0 -Force -Verbose

    ##*===============================================
    ##* POST-INSTALLATION
    ##*===============================================
    [string]$installPhase = 'Post-Installation'

    ## <Perform Post-Installation tasks here>

    ## Display a message at the end of the install
    If (-not $useDefaultMsi) { Show-InstallationPrompt -Message 'You can customize text to appear at the end of an install or remove it completely for unattended installations.' -ButtonRightText 'OK' -Icon Information -NoWait }
  }
  ElseIf ($deploymentType -ieq 'Uninstall')
  {
    ##*===============================================
    ##* PRE-UNINSTALLATION
    ##*===============================================
    [string]$installPhase = 'Pre-Uninstallation'

    ## Show Welcome Message, close required applications with a 60 second countdown before automatically closing
    Show-InstallationWelcome -CloseApps "MSACCESS,EXCEL,INFOPATH,ONENOTEM,GROOVE,ONENOTE,OUTLOOK,POWERPNT,WINPROJ,MSPUB,SPDESIGN,lync,VISIO,WINWORD,Teams,Onedrive" -AllowDeferCloseApps -AllowDefer -DeferDays "1" -CloseAppsCountdown "5400" -PersistPrompt -BlockExecution

    ## Show Progress Message (with the default message)
    Show-InstallationProgress -StatusMessage "We are removing $installTitle. Please wait." -WindowLocation 'BottomRight' -TopMost $false

    ## <Perform Pre-Uninstallation tasks here>

    ##*===============================================
    ##* UNINSTALLATION
    ##*===============================================
    [string]$installPhase = 'Uninstallation'

    ## Handle Zero-Config MSI Uninstallations
    If ($useDefaultMsi) {
      [hashtable]$ExecuteDefaultMSISplat =  @{ Action = 'Uninstall'; Path = $defaultMsiFile }; If ($defaultMstFile) { $ExecuteDefaultMSISplat.Add('Transform', $defaultMstFile) }
      Execute-MSI @ExecuteDefaultMSISplat
    }

    # <Perform Uninstallation tasks here>
    Execute-Process -Path "$dirFiles\setup.exe" -WaitForMsiExec "/configure `"$dirSupportFiles\Uninstall.xml`""

    ##*===============================================
    ##* POST-UNINSTALLATION
    ##*===============================================
    [string]$installPhase = 'Post-Uninstallation'

    ## <Perform Post-Uninstallation tasks here>

  }
  ElseIf ($deploymentType -ieq 'Repair')
  {
    ##*===============================================
    ##* PRE-REPAIR
    ##*===============================================
    [string]$installPhase = 'Pre-Repair'

    ## Show Progress Message (with the default message)
    Show-InstallationProgress -StatusMessage "We are repairing $installTitle. Please wait." -WindowLocation 'BottomRight' -TopMost $false

    ## <Perform Pre-Repair tasks here>

    ##*===============================================
    ##* REPAIR
    ##*===============================================
    [string]$installPhase = 'Repair'

    ## Handle Zero-Config MSI Repairs
    If ($useDefaultMsi) {
      [hashtable]$ExecuteDefaultMSISplat =  @{ Action = 'Repair'; Path = $defaultMsiFile; }; If ($defaultMstFile) { $ExecuteDefaultMSISplat.Add('Transform', $defaultMstFile) }
      Execute-MSI @ExecuteDefaultMSISplat
    }
    # <Perform Repair tasks here>

    ##*===============================================
    ##* POST-REPAIR
    ##*===============================================
    [string]$installPhase = 'Post-Repair'

    ## <Perform Post-Repair tasks here>

    }
  ##*===============================================
  ##* END SCRIPT BODY
  ##*===============================================

  ## Call the Exit-Script function to perform final cleanup operations
  Exit-Script -ExitCode $mainExitCode
}
Catch {
  [int32]$mainExitCode = 60001
  [string]$mainErrorMessage = "$(Resolve-Error)"
  Write-Log -Message $mainErrorMessage -Severity 3 -Source $deployAppScriptFriendlyName
  Show-DialogBox -Text $mainErrorMessage -Icon 'Stop'
  Exit-Script -ExitCode $mainExitCode
}
